tabular/src/autogluon/tabular/models/tab_transformer/tab_transformer_model.py

```diff
@@ -0,0 +1,504 @@
+
+                    best_val_epoch = e
+                    os.makedirs(os.path.dirname(self.path), exist_ok=True)
+                    torch.save(self.model, self.path + self.temp_file_name)
+
+            # If time limit has exceeded or we haven't improved in some number of epochs, stop early.
+            if e - best_val_epoch > epochs_wo_improve:
+                break
+            if time_limit:
+                time_elapsed = time.time() - start_time
+                time_left = time_limit - time_elapsed
+                if time_left <= 0:
+                    logger.log(20, "\tRan out of time, stopping training early.")
+                    break
+
+        if valloader is not None:
+            try:
+                self.model = torch.load(self.path + self.temp_file_name)
+                os.remove(self.path + self.temp_file_name)
+            except:
+                pass
+            logger.log(15, "Best model found in epoch %d" % best_val_epoch)
+
+    def _fit(self, X_train, y_train, X_val=None, y_val=None, X_unlabeled=None, time_limit=None, reporter=None, **kwargs):
+        import torch
+
+        if self.params['problem_type']==REGRESSION:
+            self.params['n_classes'] = 1
+        elif self.params['problem_type']==BINARY:
+            self.params['n_classes'] = 2
+        elif self.params['problem_type']==MULTICLASS:
+            self.params['n_classes'] = y_train.unique()
+
+        num_cols = X_train.shape[1]
+        if num_cols > self.params['max_columns']:
+            raise NotImplementedError(
+                f"This dataset has {num_cols} columns and exceeds 'max_columns' == {self.params['max_columns']}.\n"
+                f"Which is set by default to ensure the TabTransformer model will not run out of memory.\n"
+                f"If you are confident you will have enough memory, set the 'max_columns' hyperparameter higher and try again.\n")
+
+        train, val, unlab = self._tt_preprocess(X_train, X_val, X_unlabeled)
+
+        if self.problem_type == REGRESSION:
+            train.targets = torch.FloatTensor(list(y_train))
+            val.targets = torch.FloatTensor(list(y_val))
+        else:
+            train.targets = torch.LongTensor(list(y_train))
+            val.targets = torch.LongTensor(list(y_val))
+
+        trainloader, valloader, unlabloader = train.build_loader(
+            shuffle=True), val.build_loader(), unlab.build_loader() if unlab is not None else None
+
+        self.cat_feat_origin_cards = trainloader.cat_feat_origin_cards
+
+        self.get_model()
+
+        if X_unlabeled is not None:
+            self.tt_fit(unlabloader, valloader, y_val, state='pretrain', time_limit=time_limit, reporter=reporter)
+            self.tt_fit(trainloader, valloader, y_val, state='finetune', time_limit=time_limit, reporter=reporter)
+        else:
+            self.tt_fit(trainloader, valloader, y_val, time_limit=time_limit, reporter=reporter)
+
+    def _predict_proba(self, X, preprocess=False):
```

Comment:
must be updated due to #740 :

`def _predict_proba(self, X, **kwargs):`

and 
```
if preprocess:
   X = self.preprocess(X)
```

updated to:

```
X = self.preprocess(X, **kwargs)
```

Reply:
These changes still leave prediction broken. Investigating...

Reply:
Fixed. Was missing 2 extra methods in abstract_learner after rebase.